File: go/datastructures/min_heap.go
package datastructures

import "cmp"

// MinHeap implements a binary min-heap.
// Time: O(log n) insert/extract, O(1) peek
// Space: O(n)
type MinHeap[T cmp.Ordered] struct {
	heap []T
}

func NewMinHeap[T cmp.Ordered]() *MinHeap[T] {
	return &MinHeap[T]{heap: make([]T, 0)}
}

func (h *MinHeap[T]) Insert(value T) {
	h.heap = append(h.heap, value)
	h.siftUp(len(h.heap) - 1)
}

func (h *MinHeap[T]) ExtractMin() (T, bool) {
	var zero T
	if len(h.heap) == 0 {
		return zero, false
	}

	min := h.heap[0]
	last := h.heap[len(h.heap)-1]
	h.heap = h.heap[:len(h.heap)-1]

	if len(h.heap) > 0 {
		h.heap[0] = last
		h.siftDown(0)
	}

	return min, true
}

func (h *MinHeap[T]) Peek() (T, bool) {
	var zero T
	if len(h.heap) == 0 {
		return zero, false
	}
	return h.heap[0], true
}

func (h *MinHeap[T]) IsEmpty() bool {
	return len(h.heap) == 0
}

func (h *MinHeap[T]) Size() int {
	return len(h.heap)
}

func (h *MinHeap[T]) Clear() {
	h.heap = make([]T, 0)
}

func (h *MinHeap[T]) ToSlice() []T {
	result := make([]T, len(h.heap))
	copy(result, h.heap)
	return result
}

func (h *MinHeap[T]) siftUp(index int) {
	for index > 0 {
		parentIndex := (index - 1) / 2
		if h.heap[index] >= h.heap[parentIndex] {
			break
		}
		h.heap[index], h.heap[parentIndex] = h.heap[parentIndex], h.heap[index]
		index = parentIndex
	}
}

func (h *MinHeap[T]) siftDown(index int) {
	length := len(h.heap)

	for {
		leftChild := 2*index + 1
		rightChild := 2*index + 2
		smallest := index

		if leftChild < length && h.heap[leftChild] < h.heap[smallest] {
			smallest = leftChild
		}

		if rightChild < length && h.heap[rightChild] < h.heap[smallest] {
			smallest = rightChild
		}

		if smallest == index {
			break
		}

		h.heap[index], h.heap[smallest] = h.heap[smallest], h.heap[index]
		index = smallest
	}
}

File: kotlin/src/main/kotlin/datastructures/MinHeap.kt
package datastructures

/**
 * Binary min-heap implementation.
 * Time: O(log n) insert/extract, O(1) peek
 * Space: O(n)
 */
class MinHeap<T : Comparable<T>>(
    private val comparator: Comparator<T>? = null
) {
    private val heap = mutableListOf<T>()

    fun insert(value: T) {
        heap.add(value)
        siftUp(heap.lastIndex)
    }

    fun extractMin(): T? {
        if (heap.isEmpty()) return null

        val min = heap[0]
        val last = heap.removeAt(heap.lastIndex)

        if (heap.isNotEmpty()) {
            heap[0] = last
            siftDown(0)
        }

        return min
    }

    fun peek(): T? = heap.firstOrNull()

    fun isEmpty(): Boolean = heap.isEmpty()

    fun size(): Int = heap.size

    fun clear() {
        heap.clear()
    }

    fun toList(): List<T> = heap.toList()

    private fun siftUp(index: Int) {
        var i = index
        while (i > 0) {
            val parentIndex = (i - 1) / 2
            if (compare(heap[i], heap[parentIndex]) >= 0) break
            swap(i, parentIndex)
            i = parentIndex
        }
    }

    private fun siftDown(index: Int) {
        var i = index

        while (true) {
            val leftChild = 2 * i + 1
            val rightChild = 2 * i + 2
            var smallest = i

            if (leftChild < heap.size && compare(heap[leftChild], heap[smallest]) < 0) {
                smallest = leftChild
            }

            if (rightChild < heap.size && compare(heap[rightChild], heap[smallest]) < 0) {
                smallest = rightChild
            }

            if (smallest == i) break

            swap(i, smallest)
            i = smallest
        }
    }

    private fun swap(i: Int, j: Int) {
        val temp = heap[i]
        heap[i] = heap[j]
        heap[j] = temp
    }

    private fun compare(a: T, b: T): Int = comparator?.compare(a, b) ?: a.compareTo(b)
}
